lib/tools/xctest-commands.js

```diff
@@ -0,0 +1,51 @@
+import os from 'os';
+
+const DELIMETER = ' | ';
+
+const xctestCommands = {};
+
+/**
+ * Runs an xcuitest on the device or simulator
+ * @param {string} testRunnerBundleId The bundle id of the test runner
+ * @param {string} appUnderTestBundleId The bundle id of the app under test
+ * @param {string} xctestBundleId The bundle id of the xctest package
+ * @returns {SubProcess}
+ */
+xctestCommands.runXCUITest = async function runXCUITest (testRunnerBundleId, appUnderTestBundleId, xctestBundleId) {
+  const uiTestProcess = this.createSubProcess([
+    'xctest', 'run', 'ui', '--udid', this.udid, xctestBundleId, appUnderTestBundleId, testRunnerBundleId
+  ]);
+  await uiTestProcess.start(0);
+  return uiTestProcess;
+};
+
+/**
+ * Installs a xctest bundle on the device or simulator
+ * @param {string} xctestBundlePath the path of the xctest bundle
+ * @returns {string} the bundle id of xctest bundle that was installed
+ */
+xctestCommands.installXCTestBundle = async function installXCTestBundle (xctestBundlePath) {
+  const output = await this.exec(['xctest', 'install', '--udid', this.udid, xctestBundlePath]);
+  return output.replace('Installed: ');
+};
+
+/**
+ * Lists all the xctest bundles installed on a device or a simulator
+ * @returns {Array.<string>} the list of the xctest bundle ids
+ */
+xctestCommands.listXCTestBundles = async function listXCTestBundles () {
+  const output = await this.exec(['xctest', 'list', '--udid', this.udid]);
+  const ids = [];
+  for (const line of output.split(os.EOL)) {
```

Comment:
it would be nice to cover output parsing with a unit test

Reply:
`idb` has a JSON flag parameter, we should be using that.

For example, `idb list targets --json` looks like this:

```
{"name": "Apple Watch Series 2 - 38mm", "udid": "3B4B77A2-7DB8-48C3-942E-8745D40B01D4", "state": "Shutdown", "type": "simulator", "os_version": "watchOS 5.3", "architecture": "i386", "host": "", "port": 0, "is_local": false}
{"name": "Apple Watch Series 2 - 42mm", "udid": "A63F1C74-65A1-48AD-BE97-7D23DF351FD0", "state": "Shutdown", "type": "simulator", "os_version": "watchOS 5.3", "architecture": "i386", "host": "", "port": 0, "is_local": false}
{"name": "Apple Watch Series 3 - 38mm", "udid": "C4FD88DA-10EA-4E47-870D-936FD7D3E075", "state": "Shutdown", "type": "simulator", "os_version": "watchOS 5.3", "architecture": "i386", "host": "", "port": 0, "is_local": false}
{"name": "Apple Watch Series 3 - 42mm", "udid": "D1A95604-61F8-4FF4-88A6-534DD402EE87", "state": "Shutdown", "type": "simulator", "os_version": "watchOS 5.3", "architecture": "i386", "host": "", "port": 0, "is_local": false}
{"name": "Apple Watch Series 4 - 40mm", "udid": "770B82CB-AFA5-4EA6-80E4-2AED3AB982AB", "state": "Shutdown", "type": "simulator", "os_version": "watchOS 5.3", "architecture": "i386", "host": "", "port": 0, "is_local": false}
{"name": "Apple Watch Series 4 - 44mm", "udid": "FD993C2D-58D3-4F1D-933E-C2AB43A3548C", "state": "Shutdown", "type": "simulator", "os_version": "watchOS 5.3", "architecture": "i386", "host": "", "port": 0, "is_local": false}
{"name": "iPad Air (3rd generation)", "udid": "C9E2480B-FF07-427F-B712-8C9FC962DE78", "state": "Shutdown", "type": "simulator", "os_version": "iOS 12.4", "architecture": "x86_64", "host": "", "port": 0, "is_local": false}
{"name": "iPhone 5s", "udid": "663440C4-403F-493E-A863-A9086765AA06", "state": "Shutdown", "type": "simulator", "os_version": "iOS 12.4", "architecture": "x86_64", "host": "", "port": 0, "is_local": false}
{"name": "iPhone 6", "udid": "122D9587-0EC1-4FF9-9FDE-8437FDD5732F", "state": "Shutdown", "type": "simulator", "os_version": "iOS 12.4", "architecture": "x86_64", "host": "", "port": 0, "is_local": false}
{"name": "iPhone 6 Plus", "udid": "EA5621E7-23E4-45F6-9A10-CA5BBA293CDF", "state": "Shutdown", "type": "simulator", "os_version": "iOS 12.4", "architecture": "x86_64", "host": "", "port": 0, "is_local": false}
{"name": "iPhone 6s", "udid": "88CCCCA2-217D-4D36-A329-EA941F8FE119", "state": "Shutdown", "type": "simulator", "os_version": "iOS 12.4", "architecture": "x86_64", "host": "", "port": 0, "is_local": false}
{"name": "iPhone 6s Plus", "udid": "5637C2DA-1B8E-4E78-9CD5-2384C5415EF4", "state": "Shutdown", "type": "simulator", "os_version": "iOS 12.4", "architecture": "x86_64", "host": "", "port": 0, "is_local": false}
{"name": "iPhone 7", "udid": "9BDF96E2-D987-41BF-BED9-C7C19FDB1D96", "state": "Shutdown", "type": "simulator", "os_version": "iOS 12.4", "architecture": "x86_64", "host": "", "port": 0, "is_local": false}
{"name": "iPhone 7 Plus", "udid": "F909DDE9-C824-4D2E-813A-5A21B8BF0DA1", "state": "Shutdown", "type": "simulator", "os_version": "iOS 12.4", "architecture": "x86_64", "host": "", "port": 0, "is_local": false}
{"name": "iPhone 8", "udid": "D0DA1162-CED3-4125-8C6F-1142B45FBCD6", "state": "Shutdown", "type": "simulator", "os_version": "iOS 12.4", "architecture": "x86_64", "host": "", "port": 0, "is_local": false}
{"name": "iPhone 8 Plus", "udid": "926B92D9-8B47-4A9F-AE44-9F91AAA69BE5", "state": "Shutdown", "type": "simulator", "os_version": "iOS 12.4", "architecture": "x86_64", "host": "", "port": 0, "is_local": false}
{"name": "iPhone SE", "udid": "1A53602E-A872-404E-804E-062F605523C3", "state": "Shutdown", "type": "simulator", "os_version": "iOS 12.4", "architecture": "x86_64", "host": "", "port": 0, "is_local": false}
{"name": "iPhone X", "udid": "49474F10-9B89-4036-8576-4E53C161C677", "state": "Shutdown", "type": "simulator", "os_version": "iOS 12.4", "architecture": "x86_64", "host": "", "port": 0, "is_local": false}
{"name": "iPhone Xs", "udid": "8F120F8E-5CC3-4A21-B954-6B7A09940CD7", "state": "Shutdown", "type": "simulator", "os_version": "iOS 12.4", "architecture": "x86_64", "host": "", "port": 0, "is_local": false}
{"name": "iPhone Xs Max", "udid": "57F6364F-C5C6-4A9A-B7B9-D55A6D524F08", "state": "Shutdown", "type": "simulator", "os_version": "iOS 12.4", "architecture": "x86_64", "host": "", "port": 0, "is_local": false}
{"name": "iPhone X\u0280", "udid": "D57A07DC-72C3-4121-971B-927AD146ED2D", "state": "Shutdown", "type": "simulator", "os_version": "iOS 12.4", "architecture": "x86_64", "host": "", "port": 0, "is_local": false}
{"name": "iPad (5th generation)", "udid": "0428139F-1257-404E-9D50-FB6E07A8F58D", "state": "Shutdown", "type": "simulator", "os_version": "iOS 12.4", "architecture": "x86_64", "host": "", "port": 0, "is_local": false}
{"name": "iPad (6th generation)", "udid": "1453DE9C-E139-4D04-8869-B6D06844D544", "state": "Shutdown", "type": "simulator", "os_version": "iOS 12.4", "architecture": "x86_64", "host": "", "port": 0, "is_local": false}
{"name": "iPad Air", "udid": "E67F1C98-FCA9-410E-967C-506165B3F53F", "state": "Shutdown", "type": "simulator", "os_version": "iOS 12.4", "architecture": "x86_64", "host": "", "port": 0, "is_local": false}
{"name": "iPad Air 2", "udid": "5F2BD22B-15D7-43FF-AA9D-6C85CDA8CDC0", "state": "Shutdown", "type": "simulator", "os_version": "iOS 12.4", "architecture": "x86_64", "host": "", "port": 0, "is_local": false}
{"name": "iPad Pro (10.5-inch)", "udid": "C40E98CE-693E-43A5-B4DD-FF6205BCE153", "state": "Shutdown", "type": "simulator", "os_version": "iOS 12.4", "architecture": "x86_64", "host": "", "port": 0, "is_local": false}
{"name": "iPad Pro (11-inch)", "udid": "BE623EF7-9BBA-450E-952A-DBFEF7B9C81B", "state": "Shutdown", "type": "simulator", "os_version": "iOS 12.4", "architecture": "x86_64", "host": "", "port": 0, "is_local": false}
{"name": "iPad Pro (12.9-inch)", "udid": "307AA03F-D696-4D4C-B339-6A4BF9A58FC3", "state": "Shutdown", "type": "simulator", "os_version": "iOS 12.4", "architecture": "x86_64", "host": "", "port": 0, "is_local": false}
{"name": "iPad Pro (12.9-inch) (2nd generation)", "udid": "6E2D61F8-69F6-45BD-AC5D-7C0706A2C3E0", "state": "Shutdown", "type": "simulator", "os_version": "iOS 12.4", "architecture": "x86_64", "host": "", "port": 0, "is_local": false}
{"name": "iPad Pro (12.9-inch) (3rd generation)", "udid": "E09DACCE-F6C0-439B-9531-A1242CAA4D50", "state": "Shutdown", "type": "simulator", "os_version": "iOS 12.4", "architecture": "x86_64", "host": "", "port": 0, "is_local": false}
{"name": "iPad Pro (9.7-inch)", "udid": "9DBBCDDD-E662-4D39-ABC8-D60A988BEFB0", "state": "Shutdown", "type": "simulator", "os_version": "iOS 12.4", "architecture": "x86_64", "host": "", "port": 0, "is_local": false}
{"name": "Apple TV", "udid": "3FCB0462-981D-424A-8429-90C442AE1AAC", "state": "Shutdown", "type": "simulator", "os_version": "tvOS 12.4", "architecture": "x86_64", "host": "", "port": 0, "is_local": false}
{"name": "Apple TV 4K", "udid": "1237EE7C-1E83-43D1-9CD4-9B3D939DDB4C", "state": "Shutdown", "type": "simulator", "os_version": "tvOS 12.4", "architecture": "x86_64", "host": "", "port": 0, "is_local": false}
{"name": "Apple TV 4K (at 1080p)", "udid": "3D38877A-EDDD-4BB2-A3C3-E6A65FC02DBC", "state": "Shutdown", "type": "simulator", "os_version": "tvOS 12.4", "architecture": "x86_64", "host": "", "port": 0, "is_local": false}
```

Reply:
IMO the `json` flag should be used everywhere and added into the `exec` function.

Reply:
I don't think this flag is supported for all calls, but it is already used where possible

Reply:
Good catch. It is supported for this operation

Reply:
I won't use the json flag since it produces a wrong JSON. It is not showing the results as an array. I created an issue on idb for this https://github.com/facebook/idb/issues/544